File: masterClass/challenges/averagePair.js
/* given a sorted arr of ints, determine if there is a pair of values in the array where the average of the pair equals the target average. There may be more than one pair.

BONUS CONSTRAINTS:
Time O(n)
Space O(1)

averagePair([1, 2, 3], 2.5) -> true
averagePair([1, 3, 3, 5, 6, 7, 10, 12, 19], 8) -> true
averagePair([-1, 0, 3, 4, 5, 6], 4.1) -> false
averagePair([], 4) -> false

*/

const averagePair = function(numsArr, targetAv) {
  let startIdx = 0;
  let endIdx = numsArr.length - 1;

  while (startIdx < endIdx) {
    let curAv = avOfTwo(numsArr[startIdx], numsArr[endIdx]);
    if (curAv === targetAv) return true;
    if (curAv < targetAv) startIdx++;
    else endIdx--;
  }

  return false;
};

function avOfTwo(numOne, numTwo) {
  return (numOne + numTwo) / 2;
}

averagePair([], 4);

File: masterClass/py/problem_patterns/challenges/average_pair.py
import statistics

def average_pair(nums_list, target_av):
    """
    Given a sorted array of ints and a target average,
    determine if there is a pair of values in the array where the 
    average of the pair equals the target average.
    There may be more than one pair.
    """
    # the array is sorted - so start with the biggest average
    small_i = 0
    big_i = len(nums_list) - 1
    while small_i < big_i:
        cur_av = statistics.mean([nums_list[small_i], nums_list[big_i]])
        # if it's too big, move the big pointer left
        if cur_av > target_av:
            big_i -= 1
        # if it's too small, move the small pointer right
        elif cur_av < target_av:
            small_i += 1
        else:
            return True
    # when the pointers meet or cross, return False
    return False    

print(average_pair([1, 2, 3], 2.5)) # True
print(average_pair([1, 3, 3, 5, 6, 7, 10, 12, 19], 8)) # True
print(average_pair([-1, 0, 3, 4, 5, 6], 4.1)) # False